src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,122 @@
+Think of `u` as the image of `X` in an algebra map `A[X] → module.End A M`
+and extend this to all polynomials `p(X)` using the `A` algebra structure
+on `module.End A M` (`= (M →ₗ[A] M)`).  This is given to us by `polynomial.aeval`.
+Using this definition, we can define the annihilating ideal of `u` to be
+`{ p ∈ A[x] | p(u) = 0 }`. This is an ideal in `A[X]`. -/
+noncomputable def annihilating_ideal (u : M →ₗ[A] M) : ideal A[X] :=
+ring_hom.ker (aeval u).to_ring_hom
+
+/-- It is useful to refer to ideal membership sometimes
+ and the annihilation condition other times -/
+lemma mem_annihilating_ideal_iff_aeval_0 (u : M →ₗ[A] M) (p : A[X]) :
+  p ∈ annihilating_ideal u ↔ aeval u p = 0 :=
+iff.rfl
+
+variables {𝕜 V : Type*}
+variables [field 𝕜] [add_comm_group V] [module 𝕜 V]
+
+/-- Since `𝕜[x]` is a principal ideal domain there is a polynomial `g` such that
+ `span 𝕜 {g} = annihilating_ideal u` -/
+noncomputable def annihilating_ideal_generator (u : V →ₗ[𝕜] V) : 𝕜[X] :=
+submodule.is_principal.generator (annihilating_ideal u)
+
+/-- We are working toward showing the generator of the annihilating ideal
+in the field case is the minimal polynomial. We are going to use a uniqueness
+theorem of the minimal polynomial. This is the first condition: it must annihilate
+the original endomorphism `u : V →ₗ[𝕜] V`. -/
+lemma annihilating_ideal_generator_aeval_0 (u : V →ₗ[𝕜] V) :
+  aeval u (annihilating_ideal_generator u) = 0 :=
+begin
+  rw annihilating_ideal_generator,
+  have gen_u_member := submodule.is_principal.generator_mem (annihilating_ideal u),
+  exact (ring_hom.mem_ker (polynomial.aeval u).to_ring_hom).1 gen_u_member,
+end
+
+/-- This is a stepping stone to show the generator has minimal degree -/
+lemma mem_iff_generator_dvd (u : V →ₗ[𝕜] V) (p : 𝕜[X]) :
+  p ∈ annihilating_ideal u ↔ annihilating_ideal_generator u ∣ p :=
+submodule.is_principal.mem_iff_generator_dvd (annihilating_ideal u)
+
+/-- We are working toward showing the generator of the annihilating ideal
+in the field case is the minimal polynomial. We are going to use a uniqueness
+theorem of the minimal polynomial. This is the second condition: it must have minimal
+degree among the annihilators of the original endomorphism `u : V →ₗ[𝕜] V`. -/
+lemma non_zero_gen_of_non_zero_mem (u : V →ₗ[𝕜] V) (p : 𝕜[X])
+  (hp : p ∈ annihilating_ideal u) (hpn0 : p ≠ 0) :
+  annihilating_ideal_generator u ≠ 0 :=
+begin
+  intro hg,
+  cases (mem_iff_generator_dvd u p).1 hp with q hq,
+  rw [ hg, zero_mul ] at hq,
+  exact hpn0 hq,
+end
+
+/-- The generator of the annihilating ideal has minimal degree among
+ the non-zero members of the annihilating ideal -/
+lemma degree_annihilating_ideal_generator_le_of_mem (u : V →ₗ[𝕜] V) (p : 𝕜[X])
+  (hp : p ∈ annihilating_ideal u) (hpn0 : p ≠ 0) :
+  degree (annihilating_ideal_generator u) ≤ degree p :=
+begin
+  apply degree_le_of_dvd hpn0,
+  cases (mem_iff_generator_dvd u p).1 hp with q hpgq,
+  exact dvd.intro q (eq.symm hpgq),
+end
```

Comment:
There's no need to unpack then reassemble the dvd proof here:

```suggestion
degree_le_of_dvd hpn0 $ (mem_iff_generator_dvd u p).1 hp
```

Reply:
i only now understand what the dollar sign means. Before understanding it I started the proof over and got
```
degree_le_of_dvd hpn0 ((mem_iff_generator_dvd u p).1 hp)
```
to work. Is that okay, or do we try to avoid parens when possible?

Reply:
The dollar sign's mainly a convenience -- if you like the aesthetics of parentheses more you're free to use them.